e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -190,67 +196,155 @@ func (pf *PortForwarder) portForwardToPod(pod string, async bool) error {
 	}
+	pf.isStarted = true
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
-	go func() {
-		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
+	errorCh := wrapWithChannel(func() error {
+		for { // loop forever until context is done or a terminal error
+			err := pf.startAndWait()
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				pf.logger.Infof("PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+				return nil
+			} else if err != nil && isTerminalError(err) {
+				err = errors.Wrapf(err, "PortForwarder for Pod %s/%s encountered a fatal error", pf.ns, pf.pod)
 				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
-				return
+				return err
+			} else if err != nil {
+				pf.logger.Infof("port-forward must be restarted for Pod %s/%s: %v", pf.ns, pf.pod, err)
+			} else {
+				pf.logger.Infof("port-forward must be restarted for Pod %s/%s: nil error", pf.ns, pf.pod)
 			}
 		}
+	})
+	return errorCh
+}
+
+// startAndWait will start the port-forward and block until seeing an indicator
+// that the port-forward is no longer healthy. If that occurs, it will return.
+func (pf *PortForwarder) startAndWait() error {
+	defer pf.notReady()
+	if err := pf.portForwardToDeployment(); err != nil {
+		return err
+	}
+	return pf.waitForFailure()
+}
+
+// waitForFailure waits for any failure condition which indicates the port-forward
+// must be restarted. The port-forward must be restarted if:
+// - the subprocess is killed
+// - the pod is restarted
+func (pf *PortForwarder) waitForFailure() (err error) {
+	ctx, cancel := context.WithCancel(pf.ctx)
+	cmdChan := wrapWithChannel(func() error {
+		return pf.waitForCmdFailure(ctx)
+	})
+	podChan := wrapWithChannel(func() error {
+		return pf.waitForPodFailure(ctx)
+	})
+	defer func() {
+		// before we return, cancel and wait for goroutines to exit
+		cancel()
+		// If any channel returns a TerminalError, return it instead of original error.
+		// This informs upstream logic that a fatal error has occurred, and we must exit the retry loop.
+		cmdErr, ok := <-cmdChan
+		if ok && isTerminalError(cmdErr) {
```

Comment:
I think you lost the checks for Cancelled and DeadlineExceeded here. Or are they wrapped as terminal somewhere i missed?

Reply:
The check is done in the top level retry loop

Reply:
Oh, I see what's happening. I was confused because this defer is above the select.

If a terminal error is received here, it REPLACES the `err` returned by the select block. Otherwise, if a 2nd error was received, it seems to be ignored in favor of the first err (which could be nil, if Cmd.Wait returns nil).

I guess we do want to ignore Cancelled here, since one being completed will cancel the other. But I'm not sure if we want to ignore other non-terminal non-cancel errors... We probably at least want to log them?

Reply:
The intention here is just to block until any error happens which requires us to restart, i.e. a race where we don't care who wins. However if any terminal errors occur, we exit instead of restarting - hence we return the terminal error instead.

I think we could log other errors as well (errors other than cancel or terminal), although I feel it might just be unneeded noise since all we really care about is that we need to restart the port-forward

Reply:
I think we should log these, but it doesn't need to block this PR. We're gonna need all the info we can get debugging this if it fails. We can always remove noisy logs later.